### DISCUSSION TITLE
Question about PRB algorithm

### DISCUSSION THREAD

**Discussion by saeedhd96** [mitsuba-renderer/mitsuba3#581] [Q&A] [answered]
Hello,

I had a question, if we use Automatic differentation on `path` intergarator or also `prb` integrator, should they generate the exact same gradients? 

In other words, is the following statement correct? 
The only difference between these two integrators is the time and memory requirements and otherwise an optimization process using these two methods should look the exact same, given the same seed.

I get different results for optimizing materials in a room using these two integrators and I wonder why.

Thanks,
-Saeed

**Comment by njroussel** [comment, marked as answer]
Hi @saeedhd96 

I believe you are correct. The `path` and `prb` integrators should produce the exact same primal image (noise included), and their gradients should match too.

You can maybe check the gradients at a single iteration to verify this.

**Reply by njroussel** [reply]
I can confirm this is at least the expected behaviour. In section 4.1 of [Vicini et al. 2021](http://rgl.s3.eu-central-1.amazonaws.com/media/papers/Vicini2021PathReplay.pdf):

> Our method produces the same
> result as automatic differentiation, while using a constant amount
> of storage and retaining the linear time complexity of regular for-
> ward path tracing. The results do not just match conventional AD in
> expectation: they produce the same noise pattern, with only small
> numerical differences related to the different evaluation in terms of
> IEEE-754 floating point arithmetic.

The implementation might have a bug if this isn't true in practice :disappointed:

**Reply by dvicini** [reply]
@saeedhd96 Hm yes it's possible that for some reason the seeding of these two integrators isn't consistent. But generally the statement is indeed true. Are you evaluating texture gradients?

**Reply by saeedhd96** [reply]
Please see this jupyter notebook, I am showing it on a scalar roughness parameter:

https://drive.google.com/drive/folders/16-wUN6of_MgakqUWW0mefwuWZYhFNJgZ?usp=sharing

**Reply by dvicini** [reply]
I see. Forward-mode is also quite different from reverse-mode for these integrators. So its possible that somewhere in there there is a difference in how the random numbers are consumed in these integrators.
